server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/vacation/MemoryMailboxChangeRepository.java

```diff
@@ -19,47 +19,194 @@
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
 
 public class MemoryMailboxChangeRepository implements MailboxChangeRepository {
 
-    public static final int DEFAULT_NUMBER_OF_CHANGES = 10;
+    public static class MailboxChanges {
+
+        public static class MailboxChangesBuilder {
+
+            public static class MailboxChangeCollector implements Collector<MailboxChange, MailboxChanges.MailboxChangesBuilder, MailboxChanges> {
+                private int limit;
+
+                public MailboxChangeCollector(int limit) {
+                    this.limit = limit;
+                }
+
+                @Override
+                public Supplier<MailboxChanges.MailboxChangesBuilder> supplier() {
+                    return () -> new MailboxChanges.MailboxChangesBuilder(limit);
+                }
+
+                public BiConsumer<MailboxChanges.MailboxChangesBuilder, MailboxChange> accumulator() {
+                    return MailboxChanges.MailboxChangesBuilder::add;
+                }
+
+                @Override
+                public BinaryOperator<MailboxChanges.MailboxChangesBuilder> combiner() {
+                    return MailboxChanges.MailboxChangesBuilder::addAll;
+                }
+
+                @Override
+                public Function<MailboxChanges.MailboxChangesBuilder, MailboxChanges> finisher() {
+                    return MailboxChanges.MailboxChangesBuilder::build;
+                }
+
+                @Override
+                public Set<Characteristics> characteristics() {
+                    return Sets.immutableEnumSet(Characteristics.UNORDERED);
+                }
+            }
+
+            private State state;
+            private boolean hasMoreChanges;
+            private int numberOfChanges;
+            private int limit;
+            private List<MailboxId> created;
+            private List<MailboxId> updated;
+            private List<MailboxId> destroyed;
+
+            public MailboxChangesBuilder(int limit) {
+                this.limit = limit;
+                this.created = new ArrayList<>();
+                this.updated = new ArrayList<>();
+                this.destroyed = new ArrayList<>();
+            }
+
+            public MailboxChangesBuilder add(MailboxChange change) {
+                if (numberOfChanges + change.getTotalChanges() > limit) {
+                    hasMoreChanges = true;
+                    return this;
+                }
+
+                state = change.getState();
+                this.created.addAll(change.getCreated());
+                this.updated.addAll(change.getUpdated());
+                this.destroyed.addAll(change.getDestroyed());
+                this.numberOfChanges += change.getCreated().size() + change.getUpdated().size() + change.getDestroyed().size();
```

Comment:
Why should we maintain this state while we can do `this.created.size() + this.updated.size() + this.destroyed.size()` above ?